tensorboard/plugins/debugger_v2/debugger_v2_plugin_test.py

```diff
@@ -1091,62 +1092,77 @@ def testServeGraphOpInfoForOpWithInputsAndConsumers(self):
         self.assertLen(set(data["graph_ids"]), len(data["graph_ids"]))
         self.assertNotIn("graph_id", data)
         self.assertEqual(data["graph_ids"][-1], digests[op_index]["graph_id"])
-        self.assertLen(data["input_names"], 2)
-        self.assertTrue(data["input_names"][0])
-        self.assertTrue(data["input_names"][1])
+        self.assertNotIn("input_names", data)
         self.assertEqual(data["num_outputs"], 1)
+        self.assertLen(data["output_tensor_ids"], 1)
+        self.assertIsInstance(data["output_tensor_ids"][0], int)
         self.assertEqual(data["host_name"], _HOST_NAME)
         self.assertTrue(data["stack_frame_ids"])
 
         # Check input op properties.
         inputs = data["inputs"]
-        self.assertLen(inputs, len(data["input_names"]))
-        self.assertEqual(
-            inputs[0]["op_name"],
-            debug_data_multiplexer.tensor_name_to_op_name(
-                data["input_names"][0]
-            ),
-        )
-        self.assertEqual(
-            inputs[1]["op_name"],
-            debug_data_multiplexer.tensor_name_to_op_name(
-                data["input_names"][1]
-            ),
-        )
         # The two input tensors to the AddV2 op are from the same Unpack
         # (unstack) op that provides 4 outputs.
-        self.assertEqual(inputs[0]["op_type"], "Unpack")
-        self.assertEqual(inputs[1]["op_type"], "Unpack")
-        self.assertTrue(inputs[0]["input_names"])
-        self.assertTrue(inputs[1]["input_names"])
-        self.assertEqual(inputs[0]["num_outputs"], 4)
-        self.assertEqual(inputs[1]["num_outputs"], 4)
-        self.assertEqual(inputs[0]["host_name"], _HOST_NAME)
-        self.assertEqual(inputs[1]["host_name"], _HOST_NAME)
-        self.assertEqual(inputs[0]["graph_ids"], data["graph_ids"])
-        self.assertEqual(inputs[1]["graph_ids"], data["graph_ids"])
-        self.assertEqual(
-            inputs[0]["stack_frame_ids"], inputs[1]["stack_frame_ids"]
-        )
-        self.assertNotIn("inputs", inputs[0])
-        self.assertNotIn("inputs", inputs[1])
-        self.assertNotIn("consumers", inputs[0])
-        self.assertNotIn("consumers", inputs[0])
+        self.assertTrue(inputs[0]["op_name"])
+        self.assertEqual(inputs[0]["output_slot"], 0)
+        self.assertTrue(inputs[1]["op_name"])
+        self.assertEqual(inputs[1]["output_slot"], 1)
+        input0 = inputs[0]["data"]
+        input1 = inputs[1]["data"]
+        for inpt in (input0, input1):
+            self.assertEqual(inpt["op_type"], "Unpack")
+            self.assertNotIn("input_names", inpt)
+            self.assertEqual(inpt["num_outputs"], 4)
+            self.assertLen(inpt["output_tensor_ids"], 4)
+            self.assertEqual(inpt["host_name"], _HOST_NAME)
+            self.assertEqual(inpt["graph_ids"], data["graph_ids"])
+            self.assertLen(inpt["inputs"], 1)
```

Comment:
Can you construct the entire expected input and do a deep compare all at once?  This sequence of individual comparisons is very hard to follow.  Deep compare could be done with `assertDictEqual()` or `repr()` if nothing else.

Reply:
The difficulty in doing that is that some fields (like the detailed values of op_names) do not have concrete contracts. TensorFlow may change its internal op-naming algorithm any time and we'll be broken if we assert on a specific string name. In the TF2 world, op names are only meant for debugging. Examples like Lines 1120-1121 are written that way for exactly that reason.

I looked but didn't find a "wildcard" token supported by deep `assertEqual` - maybe a feature request for PyUnit.